simple/cm_data.c

```diff
@@ -0,0 +1,304 @@
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <string.h>
+#include <inttypes.h>
+#include <assert.h>
+
+#include <rdma/fi_errno.h>
+#include <rdma/fi_endpoint.h>
+#include <rdma/fi_cm.h>
+
+#include "shared.h"
+
+static char *cm_data;
+static size_t cm_data_size;
+static struct fi_eq_cm_entry *entry;
+
+static int server_setup(void)
+{
+	size_t opt_size;
+	int ret;
+
+	ret = ft_start_server();
+	if (ret)
+		return ret;
+
+	opt_size = sizeof(opt_size);
+	return fi_getopt(&pep->fid, FI_OPT_ENDPOINT, FI_OPT_CM_DATA_SIZE,
+		&cm_data_size, &opt_size);
+}
+
+static int client_setup(void)
+{
+	size_t opt_size;
+	int ret;
+
+	/* Get fabric info */
+	ret = fi_getinfo(FT_FIVERSION, opts.dst_addr, opts.dst_port, 0, hints,
+			&fi);
+	if (ret) {
+		FT_PRINTERR("fi_getinfo", ret);
+		return ret;
+	}
+
+	ret = ft_open_fabric_res();
+	if (ret)
+		return ret;
+
+	ret = ft_alloc_active_res(fi);
+	if (ret)
+		return ret;
+
+	ret = ft_init_ep();
+	if (ret)
+		return ret;
+
+	opt_size = sizeof(opt_size);
+	return fi_getopt(&ep->fid, FI_OPT_ENDPOINT, FI_OPT_CM_DATA_SIZE,
+		&cm_data_size, &opt_size);
+}
+
+static int server_connect(size_t paramlen)
+{
+	size_t expected;
+	uint32_t event;
+	int ret;
+
+	expected = paramlen + sizeof(*entry);
+
+	ret = fi_eq_sread(eq, &event, entry, expected, -1, 0);
+	if (ret != expected) {
+		FT_PROCESS_EQ_ERR(ret, eq, "fi_eq_sread", "listen");
+		return ret;
+	}
+
+	ret = ft_check_buf(entry->data, paramlen);
+	if (ret)
+		return -FI_EOTHER;
+
+	fi = entry->info;
+	if (event != FI_CONNREQ) {
+		FT_ERR("Unexpected CM event %d", event);
+		ret = -FI_EOTHER;
+		goto err;
+	}
+
+	ret = fi_domain(fabric, fi, &domain, NULL);
+	if (ret) {
+		FT_PRINTERR("fi_domain", ret);
+		goto err;
+	}
+
+	ret = ft_alloc_active_res(fi);
+	if (ret) {
+		FT_PRINTERR("alloc_active_res", ret);
+		goto err;
+	}
+
+	ret = ft_init_ep();
+	if (ret) {
+		FT_PRINTERR("init_ep", ret);
+		goto err;
+	}
+
+	ft_fill_buf(cm_data, paramlen);
+
+	/* Accept the incoming connection. Also transitions endpoint to active state */
+	ret = fi_accept(ep, cm_data, paramlen);
+	if (ret) {
+		FT_PRINTERR("fi_accept", ret);
+		goto err;
+	}
```

Comment:
I like this test.  What could make it even more useful is if the passive side handled calling both fi_accept and fi_reject -- either as separate loops, or alternating calls, or via command line option, or...